module/x/gravity/keeper/attestation_handler.go

```diff
@@ -21,317 +22,510 @@ var _ types.DistributionKeeper = (*distrkeeper.Keeper)(nil)
+	}
+}
+
+// Upon acceptance of sufficient validator SendToCosmos claims: transfer tokens to the appropriate cosmos account
+// The cosmos receiver can be a native account (e.g. gravity1abc...) or a foreign account (e.g. cosmos1abc...)
+// In the event of a native receiver, bank module handles the transfer, otherwise an IBC transfer is initiated
+// Note: Previously SendToCosmos was referred to as a bridge "Deposit", as tokens are deposited into the gravity contract
+func (a AttestationHandler) handleSendToCosmos(ctx sdk.Context, claim types.MsgSendToCosmosClaim) error {
+	invalidAddress := false
+	receiverAddress, addressErr := types.IBCAddressFromBech32(claim.CosmosReceiver)
+	if addressErr != nil {
+		invalidAddress = true
+	}
+	tokenAddress, errTokenAddress := types.NewEthAddress(claim.TokenContract)
+	ethereumSender, errEthereumSender := types.NewEthAddress(claim.EthereumSender)
+	// nil address is not possible unless the validators get together and submit
+	// a bogus event, this would create lost tokens stuck in the bridge
+	// and not accessible to anyone
+	if errTokenAddress != nil {
+		hash, _ := claim.ClaimHash()
+		a.keeper.logger(ctx).Error("Invalid token contract",
+			"cause", errTokenAddress.Error(),
+			"claim type", claim.GetType(),
+			"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
+			"nonce", fmt.Sprint(claim.GetEventNonce()),
+		)
+		return sdkerrors.Wrap(errTokenAddress, "invalid token contract on claim")
+	}
+	// likewise nil sender would have to be caused by a bogus event
+	if errEthereumSender != nil {
+		hash, _ := claim.ClaimHash()
+		a.keeper.logger(ctx).Error("Invalid ethereum sender",
+			"cause", errEthereumSender.Error(),
+			"claim type", claim.GetType(),
+			"id", types.GetAttestationKey(claim.GetEventNonce(), hash),
+			"nonce", fmt.Sprint(claim.GetEventNonce()),
+		)
+		return sdkerrors.Wrap(errTokenAddress, "invalid ethereum sender on claim")
+	}
+
+	// Block blacklisted asset transfers
+	// (these funds are unrecoverable for the blacklisted sender, they will instead be sent to community pool)
+	if a.keeper.IsOnBlacklist(ctx, *ethereumSender) {
+		invalidAddress = true
+	}
+
+	// Check if coin is Cosmos-originated asset and get denom
+	isCosmosOriginated, denom := a.keeper.ERC20ToDenomLookup(ctx, *tokenAddress)
+	coin := sdk.NewCoin(denom, claim.Amount)
+	coins := sdk.Coins{coin}
+
+	moduleAddr := a.keeper.accountKeeper.GetModuleAddress(types.ModuleName)
+	if !isCosmosOriginated { // We need to mint eth-originated coins (aka vouchers)
+		preMintBalance := a.keeper.bankKeeper.GetBalance(ctx, moduleAddr, denom)
+		// Ensure that users are not bridging an impossible amount, only 2^256 - 1 tokens can exist on ethereum
+		prevSupply := a.keeper.bankKeeper.GetSupply(ctx, denom)
+		newSupply := new(big.Int).Add(prevSupply.Amount.BigInt(), claim.Amount.BigInt())
+		if newSupply.BitLen() > 256 { // new supply overflows uint256
```

Comment:
this shouldn't be possible becuase comsos-sdk int's are limited to u256 max, but that can be changed so I suppose we should code defensively.

Reply:
This check is preserved from the original code, but github shows that poorly. It appears that we're doing this conversion to big.Int math to have an opportunity to create an error rather than a generic: `panic("Int overflow")`.